cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,288 @@
+}
+
+// Get returns the current CCM worker firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if notNil(fm) {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fm.fwCache.firewall.state.ID)
+		if err != nil {
+			klog.Errorf("failed to retrieve firewall by ID from local cache")
+		}
+		if resp != nil && resp.StatusCode == 404 {
+			klog.Infof("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	firewallList, err := allFirewallList(ctx, fm.client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve firewall from DO firewall API: %s", err)
+	}
+	formattedWorkerFirewallName := fmt.Sprintf(firewallWorkerCCMNameFormat, fm.workerFirewallName)
+	for _, fw := range firewallList {
+		if fw.Name == formattedWorkerFirewallName {
+			return &fw, nil
+		}
+	}
+	return nil, nil
+}
+
+func notNil(fm *firewallManagerOp) bool {
+	if fm.fwCache == nil || fm.fwCache.firewall == nil || fm.fwCache.firewall.state == nil {
+		return false
+	}
+	return true
+}
+
+// Set applies the given inbound rules to the CCM worker firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	defer fm.fwCache.mu.RUnlock()
+
+	targetFirewall := fm.fwCache.firewall.state
+	currentFirewall, err := fm.Get(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get firewall state: %s", err)
+	}
+	// if the current firewall state retrieved from the DO firewall API does not match,
+	// we'll need to reconcile the delta between the current firewall (DO API) and the
+	// target firewall (locally cached CCM firewall).
+	if !cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
+		// update api firewall inbound rules with the expected inbound rules.
+		fwSvc := fm.client.Firewalls
+		fr := &godo.FirewallRequest{
+			Name:          currentFirewall.Name,
+			InboundRules:  svcInboundRules,
+			OutboundRules: currentFirewall.OutboundRules,
+			DropletIDs:    currentFirewall.DropletIDs,
+			Tags:          currentFirewall.Tags,
+		}
+		currentFirewall, resp, err := fwSvc.Update(ctx, currentFirewall.ID, fr)
+		if err != nil {
+			if resp != nil && resp.StatusCode == 404 {
+				// firewall does not exist, so we should create a new one.
+				fr = &godo.FirewallRequest{
+					Name:         fmt.Sprintf(firewallWorkerCCMNameFormat, fm.workerFirewallName),
+					InboundRules: svcInboundRules,
+				}
+				currentFirewall, resp, err = fwSvc.Create(ctx, fr)
+				if err != nil {
+					if resp != nil && resp.StatusCode == 404 {
+						return fmt.Errorf("failed to create new firewall due to status not found response code")
```

Comment:
It shouldn't be possible to see a 404 when we try to _create_ a firewall. All errors from the `Create` operation can be handled equally I think.